ttl/table.go

```diff
@@ -15,27 +15,141 @@
 type PhysicalTable struct {
+	// Schema is the database name of the table
 	Schema model.CIStr
 	*model.TableInfo
+	// Partition is the partition name
+	Partition model.CIStr
 	// PartitionDef is the partition definition
 	PartitionDef *model.PartitionDefinition
 	// KeyColumns is the cluster index key columns for the table
 	KeyColumns []*model.ColumnInfo
+	// KeyColumnTypes is the types of the key columns
+	KeyColumnTypes []*types.FieldType
 	// TimeColum is the time column used for TTL
 	TimeColumn *model.ColumnInfo
 }
 
+// NewPhysicalTable create a new PhysicalTable
+func NewPhysicalTable(schema model.CIStr, tbl *model.TableInfo, partition model.CIStr) (*PhysicalTable, error) {
+	if tbl.State != model.StatePublic {
+		return nil, errors.Errorf("table '%s.%s' is not a public table", schema, tbl.Name)
+	}
+
+	ttlInfo := tbl.TTLInfo
+	if ttlInfo == nil {
+		return nil, errors.Errorf("table '%s.%s' is not a ttl table", schema, tbl.Name)
+	}
+
+	timeColumn := tbl.FindPublicColumnByName(ttlInfo.ColumnName.L)
+	if timeColumn == nil {
+		return nil, errors.Errorf("time column '%s' is not public in ttl table '%s.%s'", ttlInfo.ColumnName, schema, tbl.Name)
+	}
+
+	keyColumns, keyColumTypes, err := getTableKeyColumns(tbl)
+	if err != nil {
+		return nil, err
+	}
+
+	var partitionDef *model.PartitionDefinition
+	if tbl.Partition == nil {
+		if partition.L != "" {
+			return nil, errors.Errorf("table '%s.%s' is not a partitioned table", schema, tbl.Name)
+		}
+	} else {
+		if partition.L == "" {
+			return nil, errors.Errorf("partition name is required, table '%s.%s' is a partitioned table", schema, tbl.Name)
+		}
+
+		for i := range tbl.Partition.Definitions {
+			def := &tbl.Partition.Definitions[i]
+			if def.Name.L == partition.L {
+				partitionDef = def
+			}
+		}
+
+		if partitionDef == nil {
+			return nil, errors.Errorf("partition '%s' is not found in ttl table '%s.%s'", partition.O, schema, tbl.Name)
+		}
+	}
+
+	return &PhysicalTable{
+		Schema:         schema,
+		TableInfo:      tbl,
+		Partition:      partition,
+		PartitionDef:   partitionDef,
+		KeyColumns:     keyColumns,
+		KeyColumnTypes: keyColumTypes,
+		TimeColumn:     timeColumn,
+	}, nil
+}
+
 // ValidateKey validates a key
 func (t *PhysicalTable) ValidateKey(key []types.Datum) error {
 	if len(t.KeyColumns) != len(key) {
 		return errors.Errorf("invalid key length: %d, expected %d", len(key), len(t.KeyColumns))
 	}
 	return nil
 }
+
+// EvalExpireTime returns the expired time
+func (t *PhysicalTable) EvalExpireTime(ctx context.Context, se *Session, now time.Time) (expire time.Time, err error) {
+	tz := se.GetSessionVars().TimeZone
+
+	expireExpr := t.TTLInfo.IntervalExprStr
+	unit := ast.TimeUnitType(t.TTLInfo.IntervalTimeUnit)
+
+	var rows []chunk.Row
+	rows, err = se.ExecuteSQL(
+		ctx,
+		// FROM_UNIXTIME does not support value, so we use `FROM_UNIXTIME(0) + INTERVAL <current_ts>` to present current time
```

Comment:
"FROM_UNIXTIME does not support value"

What does this mean?

Reply:
FROM_UNIXTIME does not support negative value. The comment lost the word "negative" ...